File: agent/commands-generator/src/main/java/org/matilda/commands/collectors/CommandsCollector.java
package org.matilda.commands.collectors;

import com.google.protobuf.Message;
import org.matilda.commands.MatildaCommand;
import org.matilda.commands.exceptions.AnnotationProcessingException;
import org.matilda.commands.info.CommandInfo;
import org.matilda.commands.info.ServiceInfo;

import javax.inject.Inject;
import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.List;
import java.util.stream.Collectors;

public class CommandsCollector {

    @Inject
    Types mTypes;

    @Inject
    Elements mElements;

    @Inject
    public CommandsCollector() {}

    public List<CommandInfo> collect(ServiceInfo serviceInfo, TypeElement serviceElement) {
        return serviceElement.getEnclosedElements().stream()
                .filter(element -> element.getAnnotation(MatildaCommand.class) != null)
                .map(ExecutableElement.class::cast)
                .map(element -> collectCommand(element, serviceInfo))
                .collect(Collectors.toList());
    }

    private CommandInfo collectCommand(ExecutableElement element, ServiceInfo serviceInfo) {
        return new CommandInfo(element.getSimpleName().toString(), serviceInfo,
                getParameterType(element), getReturnType(element));
    }

    private TypeMirror getParameterType(ExecutableElement element) {
        List<? extends VariableElement> params = element.getParameters();
        if (params.size() != 1) {
            throw new AnnotationProcessingException("Can only handle commands with one parameter", element);
        }

        TypeMirror type = params.get(0).asType();
        verifyType(type, element);
        return type;
    }

    private TypeMirror getReturnType(ExecutableElement element) {
        TypeMirror type = element.getReturnType();
        verifyType(type, element);
        return type;
    }

    private void verifyType(TypeMirror type, Element element) {
        TypeMirror messageBaseType = mTypes.getDeclaredType(mElements.getTypeElement(Message.class.getCanonicalName()));
        if (!mTypes.isSubtype(type, messageBaseType)) {
            throw new AnnotationProcessingException("Paramaters and return values of services have to be " +
                    "protobuf messages!", element);
        }
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/collectors/CommandsCollector.kt
package org.matilda.commands.collectors

import androidx.room.compiler.processing.*
import org.matilda.commands.MatildaCommand
import org.matilda.commands.exceptions.AnnotationProcessingException
import org.matilda.commands.info.CommandInfo
import org.matilda.commands.info.ParameterInfo
import org.matilda.commands.info.ServiceInfo
import org.matilda.commands.types.TypeConverter
import org.matilda.commands.types.isSupported
import javax.inject.Inject

class CommandsCollector @Inject constructor() {
    @Inject
    lateinit var mTypeConverter: TypeConverter

    fun collect(serviceInfo: ServiceInfo, serviceElement: XTypeElement) =
        serviceElement.getEnclosedElements()
            .filter { it.getAnnotation(MatildaCommand::class) != null }
            .map { element -> element as XMethodElement }
            .map { collectCommand(it, serviceInfo) }

    private fun collectCommand(element: XMethodElement, serviceInfo: ServiceInfo) =
        CommandInfo(element.name, serviceInfo, getParameters(element),
            getReturnType(element), element.thrownTypes)

    private fun getParameters(element: XExecutableElement) = element.parameters.map {
        getParameter(it)
    }

    private fun getParameter(element: XVariableElement) =
        ParameterInfo(element.name, element.type).also {
            verifyType(it.type, element)
        }

    private fun getReturnType(element: XMethodElement) = element.returnType.also {
        verifyType(it, element)
    }

    private fun verifyType(type: XType, element: XElement) {
        if (!mTypeConverter.isSupported(type)) {
            throw AnnotationProcessingException("Paramaters and return values of services have to be: " +
                    mTypeConverter.supportedTypesDescription, element)
        }
    }
}
